crates/suins-indexer/src/indexer.rs

```diff
@@ -0,0 +1,219 @@
+
+                name_record_changes.push(NameRecordChange {
+                    field: name_record,
+                    field_id: object.id(),
+                });
+            }
+            // Parse subdomain wrappers and save them in our hashmap.
+            // Later, we'll save the id of the wrapper in the name record.
+            // NameRecords & their equivalent SubdomainWrappers are always created in the same PTB, so we can safely assume
+            // that the wrapper will be created on the same checkpoint as the name record and vice versa.
+            if self.is_subdomain_wrapper(object) {
+                let sub_domain: SubDomainRegistration = object.to_rust().unwrap();
+                sub_domain_wrappers.insert(
+                    sub_domain.nft.domain_name,
+                    sub_domain.id.id.bytes.to_string(),
+                );
+            };
+        }
+    }
+
+    /// For each input object, we're parsing the name record deletions
+    /// A deletion we want to track is a deleted object which is of `NameRecord` type.
+    /// Domain replacements do not count as deletions, but instead are an update to the latest state.
+    pub fn parse_name_record_deletions(
+        &self,
+        checkpoint: &CheckpointData,
+        removals: &mut Vec<String>,
+    ) {
+        // Gather all object ids that got deleted.
+        // This way, we can delete removed name records
+        // (detects burning of expired names or leaf names removal).
+        let deleted_objects: Vec<_> = checkpoint
+            .transactions
+            .iter()
+            .flat_map(|x| x.effects.all_removed_objects())
+            .map(|((id, _, _), _)| id)
+            .collect();
+
+        for input in checkpoint.input_objects() {
+            if self.is_name_record(input) && deleted_objects.contains(&input.id()) {
+                removals.push(input.id().to_string());
+            }
+        }
+    }
+
+    /// Processes a checkpoint and produces a list of `updates` and a list of `removals`
+    ///
+    /// We can then use these to execute our DB bulk insertions + bulk deletions.
+    ///
+    /// Returns
+    /// - `Vec<VerifiedDomain>`: A list of NameRecord updates for the database (including sequence number)
+    /// - `Vec<String>`: A list of IDs to be deleted from the database (`field_id` is the matching column)
+    pub fn process_checkpoint(
+        &self,
+        checkpoint: CheckpointData,
+    ) -> (Vec<VerifiedDomain>, Vec<String>) {
+        let mut name_records: Vec<NameRecordChange> = vec![];
+        let mut subdomain_wrappers: HashMap<String, String> = HashMap::new();
+        let mut removals: Vec<String> = vec![];
+
+        self.parse_name_record_changes(
+            &checkpoint.output_objects(),
+            &mut name_records,
+            &mut subdomain_wrappers,
+        );
+
+        self.parse_name_record_deletions(&checkpoint, &mut removals);
```

Comment:
wonder if this could be arranged differently because there are a lot of iterations of all transactions. That is not a problem of course, not perf or anything, but it feels a more structured organization (relying less on the checkpoint helpers) could make it clear what we are doing.
I'll talk to you offline

Reply:
If I did not miss any behavior, I think there's a single iteration on the `output_objects()`, a single iteration on the `input_objects()` and an iteration on all deleted objects coming from effects.

There might indeed be a way to do this in a single loop through all tx effects, not sure if this is going to be much cleaner!

Reply:
it's just that the iterations you mentioned, iterate over all the transactions anyway. And in not being worried at all about efficiency from a performance perspective, I was wondering how things would look if we iterated over the transactions as the outer single loop.
I am sure I am not making sense and we'll chat tomorrow!